codice/ddf
# [2.13.x] DDF-04469 Add Online Certificate Status Protocol (OCSP) support

**NOTE: This is a takeover of #4471**

#### What does this PR do?
Adds Online Certificate Status Protocol (OCSP) support to DDF. An OCSP server can be queried to verify whether a certificate is revoked or not. If an error occurs during the communication with the OCSP server. An alert will be posted to the admin console.

The workflow in the current implementation is as follows (Note that we permit login unless an explicit revocation of the cert is received, IOW be as permissible as possible):
1. Check if OCSP is enabled.
1. Gather all of the configured OCSP urls + any additional OCSP urls off of the given cert.
1. Send OCSP requests to all of the found urls.
1. Iterate through the OCSP responses. If a revoked status is found, restrict login. Else, allow login.

#### Who is reviewing it?
@blen-desta 
@brjeter 
@garrettfreibott 
@ricklarsen 
@stustison 

#### Select relevant component teams:
@codice/security 

#### How should this be tested?
1. CD into the `demoCA` directory: `cd <ddf_home>/etc/certs/demoCA`.
2. Download
[`validation.cnf.zip`](https://github.com/codice/ddf/files/3036176/validation.cnf.zip) and extract `validation.cnf` to the current directory.
5. Generate the client's private key: `openssl genrsa -out clientKey.key 1024`.
6. Generate the client's unsigned certificate: `openssl req -config validation.cnf -new -x509 -days 3650 -key clientKey.key -out clientUnsignedCert.crt`. Use defaults for all the fields.
7. Generate the client's CSR: `openssl x509 -x509toreq -in clientUnsignedCert.crt -out clientCsr.csr -signkey clientKey.key`.
8. Sign the client's CSR: `openssl ca -config validation.cnf -batch -startdate 150813080000Z -enddate 250813090000Z -keyfile private/cakey.pem -cert cacert.pem -policy policy_anything -notext -out clientCert.crt -infiles clientCsr.csr`. Pass phrase is `secret`.
9. Generate the OCSP Server's CSR and key: `openssl req -config validation.cnf -new -nodes -out ocspCsr.csr -keyout ocspKey.key`. Set the common name to your hostname.
10. Sign the OCSP Server's CSR: `openssl ca -config validation.cnf -keyfile private/cakey.pem -cert cacert.pem -in ocspCSR.csr -out ocspCert.crt`. Pass phrase is `secret`. Answer `y` to both prompts.
11. Start the OCSP Server: `openssl ocsp -index index.txt -port *:8080 -rsigner ocspCert.crt -rkey ocspKey.key -CA cacert.pem -text -out log.txt`.
12. In a separate terminal, verify that the OCSP Server can validate the client's cert: `openssl ocsp -CAfile cacert.pem -issuer cacert.pem -cert clientCert.crt -url http://127.0.0.1:8080 -resp_text -noverify`.
13. Convert the client's certificate to `.p12`: `openssl pkcs12 -export -out clientCert.p12 -inkey clientKey.key -in clientCert.crt -certfile cacert.pem`. Pass phrase is `secret`.
14. Import the `.p12` client certificate into Google Chrome. Message me for instructions.
15. Log into the DDF Admin UI using basic auth: "admin:admin". Under "System -> Configuration -> Online Certificate Status Protocol", enable `Enable validating a certificate against an OCSP server` and add `http://127.0.0.1:8080` to the OCSP server URLs.
16. Incognito log into DDF using PKI with the newly added cert. You should log in without error.
17. Log out of DDF and `CTRL + C` out of the running ocsp server.
18. In the `demoCA` directory, revoke the certificate: `openssl ca -config .\validation.cnf -cert .\cacert.pem -keyfile .\private\cakey.pem -revoke clientCert.crt`.
19. In the `demoCA` directory, re-run the ocsp server: `openssl ocsp -index index.txt -port *:8080 -rsigner ocspCert.crt -rkey ocspKey.key -CA cacert.pem -text -out log.txt`.
20. In a separate terminal, verify that the OCSP Server revoked the client's cert: `openssl ocsp -CAfile cacert.pem -issuer cacert.pem -cert clientCert.crt -url http://127.0.0.1:8080 -resp_text -noverify`.
21. Incognito log into DDF using PKI with the added cert. You should see a "Credentials Invalid" error message pop up over the `Sign In With Certificate` button.
22. Log into the DDF Admin UI using basic auth: "admin:admin". Under "System -> Configuration -> Online Certificate Status Protocol", enable `Enable validating a certificate against an OCSP server` and remove all OCSP server URLs.
23. Incognito log into DDF using PKI with the added cert. You should see the same "Credentials Invalid" error message pop up over the `Sign In With Certificate` button.

#### What are the relevant tickets?
Fixes: #4469

#### Checklist:
[x]  Documentation Updated
[]    Update / Add Threat Dragon models
[x]  Update / Add Unit Tests
[]   Update / Add Integration Tests

### Notes on Review Process
Please see Notes on Review Process for further guidance on requirements for merging and abbreviated reviews.

### Review Comment Legend:
✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist.
❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
